synthia/generators/copula.py

```diff
@@ -189,22 +207,38 @@ def generate(self, n_samples: int,
 
         return samples
 
-def compute_rank_standardized(data: xr.DataArray) -> xr.DataArray:
+def compute_rank_standardized(data: xr.DataArray, is_discrete: List[bool]) -> np.ndarray:
     """Compute per-feature percentage ranks of the data. Data is a 
     2D xarray of shape (sample, feature).
 
     Example:
        >>> # 3 samples, 2 features
        >>> data = xr.DataArray([(10,0.3), (5,0.2), (1500,0.1)])
-       >>> compute_rank_standardized(data).values
+       >>> compute_rank_standardized(data, is_discrete=[False, False])
        array([[0.5 , 0.75],
               [0.25, 0.5 ],
               [0.75, 0.25]])
     """
 
     assert data.ndim == 2, f'Input array must be 2D, given: {data.ndim}'
-    # rank() requires all data in-memory, hence compute()
-    rank = data.compute().rank(data.dims[0])
-    rank_standardized = rank / (rank.max(data.dims[0]) + 1)
-    assert rank_standardized.shape == data.shape
+    data = data.compute()
+    if not any(is_discrete):
+        rank = data.rank(data.dims[0])
+        rank = rank.values
+    else:
+        # use scipy for discrete as xarray only supports 'average' rank
+        assert len(is_discrete) == data.shape[1], f"is_discrete must have length {data.shape[1]} but is {len(is_discrete)}"
+        ranks = []
+        for i in range(data.shape[1]):
+            feature = data[:,i]
+            if is_discrete[i]:
+                feature_rank_max = rankdata(feature, method='max')
+                feature_rank_min = rankdata(feature, method='min') - 1
+                ranks.append(feature_rank_max.reshape(-1, 1))
+                ranks.append(feature_rank_min.reshape(-1, 1))
+            else:
+                feature_rank = feature.rank(feature.dims[0]).values
+                ranks.append(feature_rank.reshape(-1, 1))
+        rank = np.concatenate(ranks, axis=1)
```

Comment:
The order of the columns is different from what `pv.Vinecop()` expects. The columns should be `[max_ranks_1, ..., max_ranks_d, min_rank_1, ..., min_rank_d]`.

Reply:
I see thanks but how would this work with unordered mixed/discrete columns?

Reply:
It's always [all max ranks, all min ranks], just that for continuous variables
- corresponding min rank columns are never used (and can be omitted when passed to `pv.Vinecop()`),
- max rank = avg rank (in theory at least).

Reply:
Shouldn't columns be reordered in the case of mixed continuous/discrete columns before being passed to `pv.Vinecop()`? 
E.g. `[max_rank_cont, max_rank_disc, max_rank_cont, min_rank_disc]` - > `[max_rank_cont, max_rank_cont, max_rank_disc, min_rank_disc]`. Where `max_rank_cont` are computed using the average rank in xarray and `max_rank_disc` using scipy with `method=max`. This would allow to omit computing min_ranks for continuous variables if not required by the copula

Reply:
OK probably easier with an example. 

Let's say we have three variables (X1 discrete, X2 continuous, X3 discrete). We compute max ranks for all columns (say A1, A2, A3) and min ranks - 1 only for the discrete ones (say B1, B3). Then you can pass either [A1, A2, A3, B1, A2, B3] or [A1, A2, A3, B1, B3]. 

So min ranks for the continuous variables don't have to be computed. For continuous variables (X2) you can also take A2 as the avg rank instead of max rank.

Reply:
Thanks -- This makes sense now! A2 was the issue!